File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.js
function solution(orders, course) {
    var answer = [];
    const cnt_map = new Map(course.map((c) => [c, 0]))    
    const menu_map = {}
    for (let c of course) {
        orders.forEach((order) => {
            const menu_order = order.split("")
            menu_order.sort()
            combination(menu_order, c).forEach((comb) => {
                const comb_key = comb.join("")
                menu_map[comb_key] = (menu_map[comb_key] || 0) + 1    
                cnt_map.set(c, Math.max(cnt_map.get(c), menu_map[comb_key]))                
            })
        })
    }
    
    Object.entries(menu_map).forEach(([key, value]) => {        
        if (value >= 2 && cnt_map.get(key.length) === value) {
            answer.push(key)
        }
    })
    
    answer.sort()
        
    return answer;
}

function combination(arr, selectNumber) {
    if (selectNumber === 1) return arr.map((v) => [v])
    
    const results = []
    arr.forEach((value, index, origin) => {
        const rest = origin.slice(index + 1)
        const combs = combination(rest, selectNumber - 1)
        const attached = combs.map((c) => [value, ...c])
        results.push(...attached)
    })
    
    return results
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
def solution(orders, course):
    answer = []
    max_course = dict((c, 0) for c in course)
    menu = dict()
    # course별로 조합 구하기
    for menu_cnt in course:         
        for order in orders:            
            if len(order) < menu_cnt:
                continue
            menu_list = list(order)
            menu_list.sort()
            for comb in combinations(menu_list, menu_cnt):                
                menu_str = "".join(comb)
                if menu_str not in menu:
                    menu.setdefault(menu_str, 0)
                menu[menu_str] += 1
                max_course[menu_cnt] = max(max_course[menu_cnt], menu[menu_str])
    
    course_menu = list(k for k, v in menu.items() if v >= 2 and max_course[len(k)] == v)    
    course_menu.sort()
    
    return course_menu